src/vmm/src/builder.rs

```diff
@@ -763,39 +767,659 @@ pub mod tests {
         }
-        impl devices::legacy::ReadableFd for SerialInput {};
 
+        File::open(kernel_path).unwrap()
+    }
+
+    fn default_mmio_device_manager() -> MMIODeviceManager {
+        MMIODeviceManager::new(
+            &mut (arch::MMIO_MEM_START as u64),
+            (arch::IRQ_BASE, arch::IRQ_MAX),
+        )
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn default_portio_device_manager(event_manager: &mut EventManager) -> PortIODeviceManager {
         let read_tempfile = TempFile::new("/tmp/serial_read_").unwrap();
         let read_file = File::open(read_tempfile.as_path()).unwrap();
         let read_handle = SerialInput(read_file);
+
+        let serial =
+            setup_serial_device(event_manager, Box::new(read_handle), Box::new(io::stdout()))
+                .unwrap();
+        let i8042_reset_evfd = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        PortIODeviceManager::new(serial, i8042_reset_evfd).unwrap()
+    }
+
+    fn default_kernel_cmdline() -> Cmdline {
+        let mut kernel_cmdline = kernel::cmdline::Cmdline::new(4096);
+        kernel_cmdline.insert_str(DEFAULT_KERNEL_CMDLINE).unwrap();
+        kernel_cmdline
+    }
+
+    fn default_boot_config() -> BootConfig {
+        BootConfig {
+            cmdline: default_kernel_cmdline(),
+            kernel_file: default_kernel_file(),
+        }
+    }
+
+    fn default_vmm(event_manager: &mut EventManager) -> Vmm {
+        let guest_memory = create_guest_memory(128).unwrap();
+        let kernel_cmdline = default_kernel_cmdline();
+
+        let exit_evt = EventFd::new(libc::EFD_NONBLOCK)
+            .map_err(Error::EventFd)
+            .map_err(StartMicrovmError::Internal)
+            .unwrap();
+
+        let vm = setup_kvm_vm(&guest_memory).unwrap();
+        let mmio_device_manager = default_mmio_device_manager();
+        #[cfg(target_arch = "x86_64")]
+        let pio_device_manager = default_portio_device_manager(event_manager);
+
+        Vmm {
+            events_observer: Some(Box::new(SerialStdin::get())),
+            guest_memory,
+            kernel_cmdline,
+            vcpus_handles: Vec::new(),
+            exit_evt,
+            vm,
+            mmio_device_manager,
+            #[cfg(target_arch = "x86_64")]
+            pio_device_manager,
+            seccomp_level: SECCOMP_LEVEL_ADVANCED,
+        }
+    }
+
+    fn create_x86_64_vcpus_with_template(
+        template: Option<CpuFeaturesTemplate>,
+        vcpu_count: u8,
+    ) -> crate::Result<Vec<Vcpu>> {
         let mut event_manager = EventManager::new().expect("Unable to create EventManager");
-        setup_serial_device(
-            &mut event_manager,
-            Box::new(read_handle),
-            Box::new(io::stdout()),
+        let guest_memory = create_guest_memory(128).unwrap();
+        let mut vm = setup_kvm_vm(&guest_memory).unwrap();
+        setup_interrupt_controller(&mut vm).unwrap();
+        let vcpu_config_t2_temp = VcpuConfig {
```

Comment:
Can we rename this to `vcpu_config` or something that doesn't contain `t2`?

Reply:
Yes. It's a leftover.